Evolution/NNNN-share.md

```diff
@@ -0,0 +1,178 @@
+
+It is critical to identify that this is one algorithm in the family of algorithms for sharing values. It should not attempt to solve all behavioral requirements but instead serve a common set of them that make cohesive sense together. This proposal is not mutually exclusive to the other algorithms in the sharing family.
+
+## Detailed design
+
+It is not just likely but perhaps a certainty that other algorithms will end up needing the same concept of a buffering policy beyond just AsyncStream and the new sharing mechanism. A new type in AsyncAlgorithms will be introduced to handle this. [^BufferingPolicy]
+
+```swift
+/// A strategy that handles exhaustion of a buffer’s capacity.
+public enum BufferingPolicy: Sendable {
+  /// Continue to add to the buffer, without imposing a limit on the number
+  /// of buffered elements.
+  case unbounded
+      
+  /// When the buffer is full, discard the newly received element.
+  ///
+  /// This strategy enforces keeping at most the specified number of oldest
+  /// values.
+  case bufferingOldest(Int)
+      
+  /// When the buffer is full, discard the oldest element in the buffer.
+  ///
+  /// This strategy enforces keeping at most the specified number of newest
+  /// values.
+  case bufferingNewest(Int)
+}
+```
+
+A new extension will be added to return a concrete type representing the share algorithm. This extension will take a buffering policy to identify how the buffer will be handled when iterations do not consume at the same rate.
+
+A new AsyncSequence type will be introduced that is explicitly marked as `Sendable`. This annotation identifies to the developer that this sequence can be shared and stored. Because the type is intended to be stored it cannot be returned by the extension as a `some AsyncSequence<Element, Failure> & Sendable` since that cannot be assigned to a stored property. Additionally the type of `AsyncShareSequence`, since indented to be stored, will act as a quasi erasing-barrier to the type information of previous sequences in the chain of algorithms in that it will only hold the generic information of the `Element` and `Failure` as part of it's public interface and not the "Base" asynchronous sequence it was created from.
```

Comment:
Why do we need to erase the base async sequence here? This is different than other algorithms and I'm sure this is going to have performance implications since we most likely need to box the base async sequence in an existential. 

I think using opaque return types is the _right_ solution here. Users can store that asynchronous sequence by just putting it behind an `any AsyncSequence<Element, Failure>` themselves. This leaves the choice to the user instead of forcing the existential on everyone.